source/core/Array.cpp

```diff
@@ -1121,32 +1121,58 @@ VIREO_FUNCTION_SIGNATURET(ArrayMaxMinInternal, FindArrayMaxMinInstruction)
     AQBlock1* maxValue = minValue;
     if (len == 0)
         maxIndex = minIndex = -1;
+
+    Boolean haveSeenPositiveInfinity = false;
+    Boolean haveSeenNegativeInfinity = false;
     for (IntIndex i = 0; i < len; i++) {
-        Boolean less;
-        snippet->_p0 = arrayIn->BeginAt(i);
+        Boolean shouldUpdateMaxOrMin = false;
+        AQBlock1* currentElement = arrayIn->BeginAt(i);
+        Double currentElementValue = *((Double*)currentElement);
+        Boolean isANumber = !::isnan(currentElementValue);
+        Boolean isPositiveInfinity = ::isinf(currentElementValue) && currentElementValue > 0;
+        Boolean isNegativeInfinity = ::isinf(currentElementValue) && currentElementValue < 0;
+        snippet->_p0 = currentElement;
         snippet->_p1 = minValue;
-        snippet->_p2 = &less;
+        snippet->_p2 = &shouldUpdateMaxOrMin;
         _PROGMEM_PTR(snippet, _function)(snippet);
-        if (less) {
-            minValue = arrayIn->BeginAt(i);
+        // These additional checks are necessary because IsLT does not handle NaN and Infinity the way we want it to.
+        if ((shouldUpdateMaxOrMin && isANumber) || (isNegativeInfinity && !haveSeenNegativeInfinity)) {
+            minValue = currentElement;
             minIndex = i;
         }
+
         snippet->_p0 = maxValue;
-        snippet->_p1 = arrayIn->BeginAt(i);
-        snippet->_p2 = &less;
+        snippet->_p1 = currentElement;
+        snippet->_p2 = &shouldUpdateMaxOrMin;
         _PROGMEM_PTR(snippet, _function)(snippet);
-        if (less) {
-           maxValue = arrayIn->BeginAt(i);
-           maxIndex = i;
+        if ((shouldUpdateMaxOrMin && isANumber) || (isPositiveInfinity && !haveSeenPositiveInfinity)) {
```

Comment:
If the logic of handling nan's and infinities in this primitive is isolated to this primitive, what happens when other comparison primitives such as "<", ">" are used with nan's and infinities? Should we not fix those as well or rather should we fix those first?

Reply:
Actually, I think we have a much bigger problem in the existing code. DECLARE_VIREO_COMPARISON_PRIMITIVES_BASE and other macros (IsLTSort, IsEqSearch) that work on all types are casting double unconditionally and suffer the same issues with the proposed implementation in this PR. Those must be fixed as well. isnan, infinity should be called only for float and double and that too handled separately for those float and double themselves as the sizes are different.

I would say we should start cleaning up from the macros (start from the basic ones such as IsLT, IsGT, IsLTSort, ..) and then try to use them as comparators to fix the higher primitives (Search/Sort/ArrayMaxMin) and have as minimal logic as possible (and no logic at all if possible) in the higher primitives.

Reply:
I understand and agree with you. This will occur for the several values of the different integer types, that when cast to double will incorrectly be interpreted as infinity or nan. Because this problem has already existed I would like to submit this as it is and create another CAR to clean up the error you mention.

Reply:
Discussed with Michael about what all at the minimum we should do for this PR and what else should be tracked separately.